### ISSUE TITLE
Why asynchronous downloadOnly() requires to be called from main thread ?

### ISSUE THREAD

**Issue by tao1** [bumptech/glide#961]
Hi,

I need to cache all portrait covers of all movies of a pretty big VOD database to make them available in global Android TV search. Indeed android.app.SearchManager.SUGGEST_COLUMN_RESULT_CARD_IMAGE only accepts local images.
So I tried to use asynchronous downloadOnly() method to cache every movie cover in a SyncAdapter runnning in its dedicated thread. 
Unfortunately asynchronous downloadOnly() method requires to be called from main thread because it calls internally GenericRequestBuilder.into() method which calls Util.assertMainThread().
So I tried to call asynchronous downloadOnly() method in main thread but it leads to UI hangs. This limitation appears to me as an undesirable side effect of GenericRequestBuilder.into() call which usually takes a 'view' target.
But I am not sure of my conclusion, so I prefer to ask instead of forking your fucking good piece of software.

Kind regards,

Laurent.

**Comment by TWiStErRob** [comment]
Hmm, which version of Glide are you using? #459 has a good example of background downloadOnly... did it fail with an exception or you just read the code? Which overload of downloadOnly do you call?

**Comment by tao1** [comment]
I get the following exception using version 3.7.0:
java.lang.IllegalArgumentException: You must call this method on the main thread
            at com.bumptech.glide.util.Util.assertMainThread(Util.java:135)
            at com.bumptech.glide.GenericRequestBuilder.into(GenericRequestBuilder.java:642)
            at com.bumptech.glide.GenericTranscodeRequest.downloadOnly(GenericTranscodeRequest.java:89)
            at com.bumptech.glide.DrawableTypeRequest.downloadOnly(DrawableTypeRequest.java:96)
            at fr.bouyguestelecom.tv.vod.data.CacheCoversTask.doInBackground(CacheCoversTask.java:55)

Your provided sample use synchronous downloadOnly() method which is not very convenient for a bunch of downloads. Indeed to wait for all downloads to complete it is more elegant to use a SimpleTarget<File> as a callback instead of waiting for every FutureTarget returned by synchronous downloadOnly() call.
I use a java.util.concurrent.CountLatchDown in callbacks and just wait for counter reaches zero in my main loading thread.
If my explanation is not clear, I can elaborate more.
BTW thank you for answering.

**Comment by TWiStErRob** [comment]
The reason why it doesn't work for you is right after the sync/async "definition":

> To download images on a background thread, you must use the synchronous version.

Here's why you **must**: `Target.onResourceReady` is called on the UI thread, and not the thread that started the load. So if you actually do something with the received File you would be blocking the UI.

It's really easy to just fire up an executor, submit all your runnables or callables that wrap the sync usage of Glide and optionally return the File (in case you want a list of them in the end). You can also wait for all of them to finish without an extra latch. With this approach you can even replace Glide with something else.

There may even be an API to just simply simply call get() on a list of Futures and return a list of what they returned, so you only need one thread (the workers are hidden inside Glide's executor service).

As a total 180: why don't you just download a zip that contains all the images? It would be much less intensive work on both the server and client. You can write server code that returns all the new images since a given date to update local cache (with default value of `Date.MIN_VALUE`). Note that a lot depends on how you're planning to update your local mirror inside Glide's cache. It's also good that you know that ages might be evicted, so if your other API requires local files, real hand-managed local files may be better.

**Comment by tao1** [comment]
Oh sorry for my no reply.
As I already said, synchronous methods are not so convenient to load a bunch of images in background. So I patched Glide for it takes a Looper as initialization argument.
Here is my patch. It's dirty but it works for me.
[Glide.with-Context-Looper-method-added.txt](https://github.com/bumptech/glide/files/180753/Glide.with-Context-Looper-method-added.txt)

**Comment by TWiStErRob** [comment]
Wow, that's a lot of code @sjudd you might want to check that "PR" out.

Here's how I would've done it. I'm really curious why you say this solution is not convenient?
```java
new Downloader(Glide.with(getApplicationContext())).execute("url","url2","url3");

class Downloader extends AsyncTask<String, File, Downloader.Result> {
	private static final String TAG = "Downloader";
	private final RequestManager glide;
	public Downloader(RequestManager glide) {
		this.glide = glide;
	}

	@Override protected Result doInBackground(String... params) {
		@SuppressWarnings("unchecked")
		FutureTarget<File>[] requests = new FutureTarget[params.length];
		// fire everything into Glide queue
		for (int i = 0; i < params.length; i++) {
			requests[i] = glide
					.load(params[i])
					.downloadOnly(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
			;
		}
		// wait for each item
		Result result = new Result();
		for (int i = 0; i < requests.length; i++) {
			try {
				result.success.put(params[i], requests[i].get(10, TimeUnit.SECONDS));
			} catch (Exception e) {
				result.failures.put(params[i], e);
			} finally {
				Glide.clear(requests[i]);
			}
		}
		return result;
	}

	@Override protected void onProgressUpdate(File... values) {
		for (File file : values) {
			Log.v(TAG, "Finished " + file);
		}
	}

	@Override protected void onPostExecute(Result result) {
		Log.i(TAG, String.format(Locale.ROOT, "Downloaded %d files, %d failed.",
				result.success.size(), result.failures.size()));
	}

	static class Result {
		Map<String, File> success = new HashMap<>();
		Map<String, Exception> failures = new HashMap<>();
	}
}
```

**Comment by tao1** [comment]
My reluctance comes from the line: 
requests[i].get(10, TimeUnit.SECONDS);

Because my "Downloader" task triggers several thousands of downloads I doubt every single download can be achieved within 10 seconds. Even if I can increase that timeout (e.g. infinite) to make it work, I disliked that approach because of the double loop: first one to trigger download, second one to get the result. Because it supposes the order should be roughly the same for the 2 loops. 

So I did prefer the callback approach simply decreasing a java.util.concurrent.CountLatchDown.

At the end you're right, my patch is pretty huge for the intended goal. But it's just your own code with the android.os.Looper addition, and the main thread assertion removal.

**Comment by TWiStErRob** [comment]
The second loop will finish in the same amount of time as a CountDownLatch would. Glide will start the loads in order (because they were queued in that order and they have the same priority). Then the second loop will start finishing the one-by-one. If you insert your `save(params[i], file)` logic right after the `file`'s declaration it will still finish in the same amount of time, because Glide is pumping the requests through on 2-8 threads, but they're an order of magnitude slower so the saving is not the bottleneck. This means that even if a single request blocks for 10 seconds, in that 10 seconds another 100 may have finished, and after the loop jumps through that hiccup it'll catch up fast because the following `.get()`s will return immediately. Also that 10 seconds is arbitrary, and likely will time out earlier even if you use Infinity because Glide's networking uses 2 seconds for network timeouts I think.

It's also possible to adapt the unpatched Glide implementation to your countDown approach, currently you have something like with the Looper patch:
```java
for() {
    Glide....downloadOnly(SimpleTarget.onResourceReady -> { save(resource); countDown(); });
}
latch.wait0();
```
you can invert that into
```java
for() {
    executorService.submit(Runnable.run -> { save(Glide....downloadOnly().get()); countDown(); }
}
latch.wait0();
```
with the benefits of latch and no custom build needed.

This inversion can be applied to any async/sync call. It's worth mentioning that Glide is not meant to be a networking library as you're essentially just downloading files in a really convoluted way (Glide's downloadOnly is meant to pre-fill the SOURCE cache); you could simply replace Glide.downloadOnly with an OkHttp or Volley request. You may be also wasting space by saving the File to another location unless you clear the disk cache upon completion or delete individual files. Saving is fine if you delete the file and then load that saved File later through `load(file).diskCacheStrategy(NONE)` to prevent duplication.

*Anyway, thanks for sharing your thoughts on the inconvenience of the double-for approach, it's always interesting to look into other people's head :) I hope my rambling makes sense above. If you decide to make any changes don't hesitate to ask questions if you need to.*

**Comment by tao1** [comment]
Your double-for approach definitely works. But I want to avoid to save a file on UI thread, or to save them sequentially after each Future.get() in my Downloader, or to submit a new thread (or at least a new task) just to wait for yours finishes because of the number of downloads: several thousands. But I am not only downloading files, I save image thumbnail permanently for android TV search, whereas Glide cache is filled with original image for my UI.
In short, I do want a callback for asynchronous operations. Moreover I do love hacking.

Thank you for your detailed replies and advices.
Cheers.

**Comment by ErikMedina** [comment]
Hi @tao1,

I'm developing the global search too and I have to cache images (and not dowload them because they are many images) to show them on cover movies. I get to cache them following [#459](https://github.com/bumptech/glide/issues/459) and pass the uri to SUGGEST_COLUMN_RESULT_CARD_IMAGE. But I have the next problem: cache images, pass uri's and return the cursor in query() takes too long that global search sometimes doesn't show the results from my app or show fewer results than existing results. Have you had a similar problem? Could you give me any suggestions on how you approach to show cached images in global search?

Thanks in advance.

**Comment by tao1** [comment]
Hi Eric,

My own use case was simpler there was a limited number of images to cache. However regarding your use case, I would recommend to return immediately when query() and start image caching. Then wait for image cache completes only in openFileDescriptor() in your ContentProvider.
